feat: folder-specific file exclusion rules

carlosedp opened the pull request:
This pull request introduces a file exclude system to the file manager, adding support for custom ignore patterns per path. It enables users to control which files and directories are hidden from view. The changes span configuration, core logic, and integration with the actor and file management systems. The PR also contemplates the API to support rules from Lua plugins which can be used to provide exclusions like gitignores.

Configuration and Integration

New [files] configuration to include `excludes` settings, allowing users to specify custom ignore patterns.

Sample configuration:

```toml
[files]
excludes = [
  # SFTP temporary files
  { urn = "*.tmp", in = "sftp://**" },
  # Python cache in search results
  { urn = "/root/**/*.pyc", in = "search://**" },
  # Multiple patterns for /code directory (supports arrays)
  { urn = [".git", ".DS_Store", "__pycache__"], in = "/code/**" },
  # Negation: show target/ even if previously ignored
  { urn = "!target/", in = "*" },
  # Ignore debug/ only inside target/ directories
  { urn = "debug", in = "/target/**" },
  #Fallback rule for all contexts
  { urn = ".DS_Store", in = "*" }
]
```

The config also adds a `excluded` command which allows to toggle, hide and show the excluded files.

Add to your `keymap.toml`:

```toml
[[mgr.prepend_keymap]]
on   = "<A-e>"
run  = "excluded toggle"
desc = "Toggle excluded files visibility"
```

The `excluded` command accepts `toggle`, `show` and `hide`.

These changes collectively add robust and flexible ignore filtering, enhancing the file manager's ability to hide unwanted files and directories according to user preferences and project standards as explained in https://github.com/sxyazi/yazi/pull/3257#issuecomment-3417508493.

There are companion PRs and repos to this PR:

- [ ] Documentation for the feature: https://github.com/yazi-rs/yazi-rs.github.io/pull/294
- [ ] Schema config update: https://github.com/yazi-rs/schemas/pull/53
- [ ] Gitignore lua plugin: https://github.com/carlosedp/gitignore.yazi

carlosedp commented:
The companion documentation PR has been updated at https://github.com/yazi-rs/yazi-rs.github.io/pull/294

carlosedp commented:
Hi @sxyazi ... any news on this? :)
Also I've been building the latest versions and it doesn't work with the Dracula flavor I'm using on the release version. Is this expected?

sxyazi commented on yazi-config/preset/yazi-default.toml:
```diff
@@ -16,82 +16,128 @@ scrolloff      = 5
 mouse_events   = [ "click", "scroll" ]
```
Please keep the file the original style

carlosedp commented:
Just pushed the changes.

carlosedp commented:
Hold the review a bit. I'll push a commit later today which adds API to allow the use of exclusion by plugins and also the lua plugin to load gitignore files from repos.

carlosedp commented:
@and-rs I override my config with yours and here it's working fine:

https://github.com/user-attachments/assets/2eb2934f-befe-4bc6-b72c-675f17ba4744

I have git.yazi too and it's not working with the upstream main branch (not related to my changes)... I've noticed many other issues with nightly like some flavors (Dracula). Yatline plugin is broken too. Maybe @sxyazi can chime in about this and if there's a guide on deprecations so plugin owners fix them.

and-rs commented:
I'll take a look at it later today. It might be the cache or maybe nested .gitignores like the repo I testing it in? Idk but I'll be trying it out soon. Ty

carlosedp commented:
No error here... even on the tiramisu repo, the path is shown correctly. I've added/removed exclude patterns and no changes (except it shows .git when removed).

Have you also updated the gitignore.yazi plugin in your plugins folder?

and-rs commented:
thanks for the fix, I haven't had issues with the keymap btw so that's great.

although I find it weird that you couldn't replicate the other bug. I did some more testing and found out it's also related with show_hidden option. I have the latest commit on both the plugin and yazi

https://github.com/user-attachments/assets/ba66bf51-3f07-470b-bdca-91eb33564a4a

sxyazi commented:
Hi @carlosedp, just wanted to let you know I haven't forgotten this PR.
It's a large feature and its implementation is more complex than I
originally expected and digging into how it works requires more time and
effort than I've had available. Right now all my available bandwidth is
tied up preparing the next release - the main branch has accumulated too
many changes and bug fixes, and I want to cut a new release soon. I should
be able to start reviewing it after that release.

On Mon, Nov 24, 2025 at 9:53 PM Carlos Eduardo ***@***.***>
wrote:

> *carlosedp* left a comment (sxyazi/yazi#3266)
> <https://github.com/sxyazi/yazi/pull/3266#issuecomment-3570898790>
>
> Hi @sxyazi <https://github.com/sxyazi> ... it's been a month since this
> is open and feels like it works fine in my day-to-day use here. Any status
> on merging this or are there any changes needed? I can merge the conflicts
> if it's good-to-go.
> Thanks!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/sxyazi/yazi/pull/3266#issuecomment-3570898790>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEFWFIDPDLYQ63JOZMGBSKT36MEW5AVCNFSM6AAAAACJW2BN42VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTKNZQHA4TQNZZGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

carlosedp commented:
Happy new year, just updated the branch with the latest merge from upstream main.

carlosedp commented:
I'm sorry but I won't be able to keep these PRs anymore. I've been busy with other stuff and I've spent more time into maintaining the PRs than using Yazi itself.

I'd love if someone could keep up and carry this on but for the time being, I'll merge them into my own branch and built it myself.

Hope someday these feature (the flat background and exclude files) can be upstreamed.
